File: simple-rest-python-flask/opensearch_web_appender.py
"""
OpenSearch Web Appender — Flask용 (HTTP 요청 추적 포함)
표준 라이브러리만 사용, 추가 의존성 없음
"""

import base64
import json
import os
import queue
import socket
import ssl
import threading
import time
import uuid
import urllib.request
from datetime import datetime, timezone, timedelta

KST = timezone(timedelta(hours=9), 'KST')


class OpenSearchWebAppender:
    def __init__(
        self,
        scheme='https',
        host='localhost',
        port=9200,
        username='',
        password='',
        app='app',
        env='local',
        max_batch_bytes=1_000_000,
        flush_interval_seconds=1,
        queue_size=8192,
    ):
        self._index       = f'logs-{app}'
        self._app         = app
        self._env         = env
        self._instance_id = os.environ.get('HOSTNAME') or socket.gethostname()
        self._max_bytes   = max_batch_bytes
        self._url         = f'{scheme}://{host}:{port}/_bulk'
        self._queue       = queue.Queue(maxsize=queue_size)
        self._auth        = None
        if username:
            creds      = base64.b64encode(f'{username}:{password}'.encode()).decode()
            self._auth = f'Basic {creds}'
        self._ssl_ctx                  = ssl.create_default_context()
        self._ssl_ctx.check_hostname   = False
        self._ssl_ctx.verify_mode      = ssl.CERT_NONE
        self._running = True
        self._thread  = threading.Thread(target=self._run, args=(flush_interval_seconds,), daemon=True)
        self._thread.start()

    def log(self, level: str, message: str, **extra):
        now   = datetime.now(timezone.utc)
        today = now.strftime('%Y.%m.%d')
        doc   = {
            '@timestamp':       now.isoformat(),
            '@timestamp_local': datetime.now().astimezone().isoformat(),
            '@timestamp_kst':   now.astimezone(KST).isoformat(),
            'level':       level,
            'message':     message,
            'app':         self._app,
            'env':         self._env,
            'instance_id': self._instance_id,
            **extra,
        }
        try:
            self._queue.put_nowait((f'{self._index}-{today}', doc))
        except queue.Full:
            pass

    def before_request(self, request):
        """Flask before_request 훅에서 호출 — trace_id, start_time 주입"""
        request.trace_id  = request.headers.get('X-Request-ID') or str(uuid.uuid4())
        request.start_time = time.time()

    def after_request(self, request, response):
        """Flask after_request 훅에서 호출 — 접근 로그 기록"""
        duration_ms = int((time.time() - request.start_time) * 1000)
        ts = datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]
        print(f'[{ts}] {request.method} {request.path} → {response.status_code} ({duration_ms}ms) [{request.trace_id}]', flush=True)
        self.log(
            'INFO',
            f'{request.method} {request.path} → {response.status_code}',
            trace_id    = request.trace_id,
            http_method = request.method,
            http_path   = request.path,
            client_ip   = request.remote_addr,
            http_status = response.status_code,
            duration_ms = duration_ms,
        )
        response.headers['X-Request-ID'] = request.trace_id
        return response

    def stop(self):
        self._running = False
        self._flush()

    def _run(self, interval: int):
        while self._running:
            time.sleep(interval)
            self._flush()

    def _flush(self):
        items = []
        try:
            while True:
                items.append(self._queue.get_nowait())
        except queue.Empty:
            pass
        if not items:
            return
        bulk = ''
        for idx, doc in items:
            bulk += json.dumps({'index': {'_index': idx}}) + '\n'
            bulk += json.dumps(doc) + '\n'
            if len(bulk.encode()) >= self._max_bytes:
                self._send(bulk)
                bulk = ''
        if bulk:
            self._send(bulk)

    def _send(self, body: str):
        try:
            data = body.encode('utf-8')
            req  = urllib.request.Request(self._url, data=data, method='POST')
            req.add_header('Content-Type', 'application/json; charset=UTF-8')
            if self._auth:
                req.add_header('Authorization', self._auth)
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=10):
                pass
        except Exception as e:
            print(f'[OpenSearch] 전송 실패: {e}', flush=True)

File: simple-rest-node-express/src/opensearch-web-appender.js
'use strict';
const https = require('https');
const http  = require('http');
const os    = require('os');
const { randomUUID } = require('crypto');

const KST_OFFSET = 9 * 60;

function _toOffsetIso(date, offsetMinutes) {
  const shifted = new Date(date.getTime() + offsetMinutes * 60000);
  const sign = offsetMinutes >= 0 ? '+' : '-';
  const hh   = String(Math.floor(Math.abs(offsetMinutes) / 60)).padStart(2, '0');
  const mm   = String(Math.abs(offsetMinutes) % 60).padStart(2, '0');
  return shifted.toISOString().replace('Z', `${sign}${hh}:${mm}`);
}

/**
 * OpenSearch _bulk API 웹 어펜더 (REST API 서비스용)
 * - HTTP 요청마다 trace_id · http_method · http_path · client_ip · http_status · duration_ms 추적
 * - 요청 완료 시 access log 자동 발행
 * - 추가 의존성 없음 (Node.js 내장 모듈만 사용)
 */
class OpenSearchWebAppender {
  constructor({
    scheme               = 'https',
    host                 = 'localhost',
    port                 = 9200,
    username             = '',
    password             = '',
    app                  = 'app',
    env                  = 'local',
    maxBatchBytes        = 1_000_000,
    flushIntervalSeconds = 1,
    queueSize            = 8192,
  } = {}) {
    this._index      = `logs-${app}`;
    this._app        = app;
    this._env        = env;
    this._instanceId = process.env.HOSTNAME || os.hostname();
    this._maxBytes   = maxBatchBytes;
    this._maxQueue   = queueSize;
    this._queue      = [];
    this._auth       = username
      ? Buffer.from(`${username}:${password}`).toString('base64')
      : null;
    this._lib    = scheme === 'https' ? https : http;
    this._parsed = new URL(`${scheme}://${host}:${port}/_bulk`);
    this._timer  = setInterval(() => this._flush(), flushIntervalSeconds * 1000);
    this._timer.unref();
  }

  log(level, message, extra = {}) {
    if (this._queue.length >= this._maxQueue) {
      console.warn(`[OpenSearch] 큐 포화 — 로그 유실 (현재: ${this._queue.length}/${this._maxQueue})`);
      return;
    }
    const now   = new Date();
    const today = now.toISOString().slice(0, 10).replace(/-/g, '.');
    const localOffset = -now.getTimezoneOffset();
    this._queue.push({
      index: `${this._index}-${today}`,
      doc: {
        '@timestamp':       now.toISOString(),
        '@timestamp_local': _toOffsetIso(now, localOffset),
        '@timestamp_kst':   _toOffsetIso(now, KST_OFFSET),
        level,
        message,
        app:         this._app,
        env:         this._env,
        instance_id: this._instanceId,
        ...extra,
      },
    });
  }

  /** Express 미들웨어 — req/res 에 traceId 주입, 응답 후 access log 발행 */
  middleware() {
    return (req, res, next) => {
      const traceId = req.headers['x-request-id'] || randomUUID();
      req.traceId = traceId;
      res.setHeader('X-Request-ID', traceId);

      const start = Date.now();
      res.on('finish', () => {
        const duration = Date.now() - start;
        const ts = new Date().toISOString().replace('T', ' ').slice(0, 23);
        console.log(`[${ts}] ${req.method} ${req.path || req.url} → ${res.statusCode} (${duration}ms) [${traceId}]`);
        this.log('INFO', `${req.method} ${req.path} → ${res.statusCode}`, {
          trace_id:    traceId,
          http_method: req.method,
          http_path:   req.path || req.url,
          client_ip:   (req.headers['x-forwarded-for'] || req.socket.remoteAddress || '').split(',')[0].trim(),
          http_status: res.statusCode,
          duration_ms: duration,
        });
      });
      next();
    };
  }

  stop() {
    clearInterval(this._timer);
    this._flush();
  }

  _flush() {
    if (this._queue.length === 0) return;
    const items = this._queue.splice(0);
    let bulk = '';
    let batchItems = [];
    for (const item of items) {
      const meta    = JSON.stringify({ index: { _index: item.index } }) + '\n';
      const docLine = JSON.stringify(item.doc) + '\n';
      const addition = Buffer.byteLength(meta + docLine);
      // 추가 전 초과 여부 확인 — 배치에 내용이 있을 때만 먼저 전송
      if (bulk && Buffer.byteLength(bulk) + addition > this._maxBytes) {
        this._send(bulk, batchItems);
        bulk = '';
        batchItems = [];
      }
      bulk += meta + docLine;
      batchItems.push(item);
    }
    if (bulk) this._send(bulk, batchItems);
  }

  _send(body, items) {
    const buf  = Buffer.from(body, 'utf8');
    const opts = {
      hostname: this._parsed.hostname,
      port:     this._parsed.port,
      path:     this._parsed.pathname,
      method:   'POST',
      headers: {
        'Content-Type':   'application/json; charset=UTF-8',
        'Content-Length': buf.length,
      },
      rejectUnauthorized: false,
    };
    if (this._auth) opts.headers['Authorization'] = `Basic ${this._auth}`;
    const req = this._lib.request(opts, (res) => { res.resume(); });
    req.on('error', (err) => {
      if (this._queue.length + items.length <= this._maxQueue) {
        this._queue.unshift(...items);
        console.warn(`[OpenSearch] 전송 실패, ${items.length}건 재큐: ${err.message}`);
      } else {
        console.error(`[OpenSearch] 전송 실패, ${items.length}건 유실 (큐 포화): ${err.message}`);
      }
    });
    req.write(buf);
    req.end();
  }
}

module.exports = OpenSearchWebAppender;
